### ISSUE TITLE
USB ReadFile() times out

### ISSUE THREAD

**Issue by zz5360** [klasyc/ScpiNet#3]
USB ReadFile times out, read with VISA can be normal, how to solve it

**Comment by klasyc** [comment]
Hi,

you did not provide much information, so it is hard to say what's going on. What instrument are you talking to? I have tested my library only with instruments that I have, which is only a few devices...

Usually, when the device does not reply at all (which leads to the ReadFile timeout), it means that the request sent by previous WriteFile function has been somehow malformed, the device just ignored it and as a result, it does not have anything useful to send back. Unfortunately, different manufacturers understand the VISA standard differently and it is very probable that my library will not work with all devices. This was the main reason, why I open sourced the library - I realized that I don't have enough devices and knowledge to make a mature library on my own.

In order to troubleshoot your problem, I would use the free ApiMonitor tool to record the ReadFile, WriteFile, ReadFileEx, and WriteFileEx function calls which makes the VISA application that works for you. Then you can compare, what sends ScpiNet instead, and finally you will be able to fix the ScpiNet library. Unfortunately, it's a tedious work. When I was adding support for my Keysight multimeter, it took me several hour to figure out, that Keysight simply does not accept Visa frames with zero sequence number, while Tektronix have no problem with this :) I think, you are facing a similar stupid issue...

**Comment by klasyc** [comment]
Hi,

I released a new version of the library which possibly could fix your issue. You can test it, if you want.

**Comment by zz5360** [comment]
I have resolved the issue. When performing the Read operation, it is necessary to follow the TMC protocol and set the length to 1024. Thank you for your response.

**Comment by zz5360** [comment]
When data == null in the CreateTmcRequest method, set len = 1024 .

**Comment by zz5360** [comment]
I noticed that the code you modified uses the same solution.

**Comment by klasyc** [comment]
This is exactly what I have fixed.

By the way, where you got the number 1024 from? It works for my Tektronix oscilloscopes, but my Keysight multimeter does not like it and I had to read only 128 bytes at a time...

**Comment by zz5360** [comment]
I went and read the USB TMC protocol documentation, and it is clearly stated in the document.

**Comment by klasyc** [comment]
And could you give me a link to that document? I have read several documents, usually of poor quality, and I haven't noticed this information anywhere...

**Comment by Elmue** [comment]
Both values are wrong.
Neither 128 Byte nor 1024 are ANY standard.
You are working with devices that have a buggy firmware.
And you are trying to adapt this class to your buggy firmware.
The size of the requested response can be anything.
My oscilloscope can transfer an entire signal over USB with the :WAVEFORM command.
The documentation says that up 250.000 bytes can  be transferred at once.
This works perfectly because they are sent in multiple USB Bulk OUT transfer packets of 512 bytes.
The oscilloscope keeps sending until all bytes are sent.
On Windows they are read into one huge buffer.
There is no limit to the transferred bytes.
You are both working with devices of poor quality.

I don't even understand why your devices care about the requested size at all ?
The size that you put into the TMC structure is the size of your receive buffer and tells the device that you cannot receive more than this limit of bytes. So if you send a request with buffersize = 1024 and the multimeter sends a small response string like "12.598\n", what is the problem? If the device sends only 7 bytes why does it care if you have a receive buffer of 128 byte or 1024 byte or hundred megabyte? I see that the engineers who developed your devices have never read the TMC documents or they did not understand them.

**Comment by klasyc** [comment]
Yes, you are right. The devices I work with have many bugs in the firmware. I also don't understand, why providing of the large buffer breaks the communication... Unfortunately, I cannot switch the devices and therefore I have to bend the drivers although I don't like it...

**Comment by klasyc** [comment]
I added a new property `DefaultBufferSize` to the `IScpiConnection` interface. This allows to tweak the (reading) buffer size and work around the buggy devices. This property can be controlled from the particular device drivers inherited from the `ScpiDevice` class, hence each driver can set the buffer size matching the needs of the controlled instrument.
